src/emailcontactstorage.cpp

```diff
@@ -0,0 +1,109 @@
+#include <algorithm>
+#include <QMailStore>
+#include "emailcontactstorage.h"
+
+Q_GLOBAL_STATIC(EmailContactStorage, storage)
+EmailContactStorage* EmailContactStorage::instance()
+{
+    if (storage != nullptr && storage->m_isInit == false)
+        storage->initialize();
+    return storage;
+}
+
+
+void EmailContactStorage::initialize()
+{
+    QMailStore *mailStore = QMailStore::instance();
+    Q_ASSERT(mailStore);
+
+    if (QMailStore::initializationState() != QMailStore::Initialized)
+        return;
+
+    QMailMessageKey incomingKey(QMailMessageKey::status(QMailMessage::Incoming));
+    QMailMessageIdList incomingIds = mailStore->queryMessages(incomingKey);
+    for (const auto& id: incomingIds) {
+        const QMailMessageMetaData metaData = mailStore->messageMetaData(id);
+        if (metaData.messageType() != QMailMessageMetaDataFwd::MessageType::Email)
+            continue;
+
+        const QMailTimeStamp timeStamp = metaData.date();
+        const QMailAddress from = metaData.from();
+        add(from, timeStamp);
+    }
+
+    QMailMessageKey outgoingKey(QMailMessageKey::status(QMailMessage::Outgoing));
+    QMailMessageIdList outgoingIds = mailStore->queryMessages(outgoingKey);
+    for (const auto& id: outgoingIds) {
+        const QMailMessageMetaData metaData = mailStore->messageMetaData(id);
+        if (metaData.messageType() != QMailMessageMetaDataFwd::MessageType::Email)
+            continue;
+
+        const QMailTimeStamp timeStamp = metaData.date();
+        const QList<QMailAddress> recipients = metaData.recipients();
+        for (const auto& address : recipients)
+            add(address, timeStamp);
+    }
+}
+
+
+int EmailContactStorage::count() const
+{
+    return m_storage.size();
+}
+
+
+QString EmailContactStorage::name(const int index) const
+{
+    if (index < 0 || index >= count())
+        return QString();
+    return m_storage[index].name();
+}
+
+
+EmailContact EmailContactStorage::at(const int index) const
+{
+    if (index < 0 || index >= count())
+        return EmailContact();
+    return m_storage[index];
+}
+
+// name can't have duplicate
+void EmailContactStorage::remove(const QString &name)
```

Comment:
You've probably already answered this elsewhere, if so I apologise: what is the name here?  Is it some combination of the display name plus email address?  (I assume it must be, otherwise two contacts with the same name but different email addresses would not be able to be represented in the storage, e.g. "chris.adams@jolla.com" / "chris.adams@jollamobile.com")

Reply:
no, this is the name (first part of the contact)
remove additional function.  Maybe with a bad name

It is needed to work with other models.   To remove what other models showed.

The phonbooks model is like that.
It is necessary that this model does not have contacts of another model. This function is just for this